---
title: "Analysis of mutants"
author: "AG"
date: "last update: `r format(Sys.Date(), format = '%d %B %Y')`"
output:
  html_document:
    toc: yes
    df_print: paged
    code_folding: hide
    theme: cerulean
    highlight: kate
    toc_float: yes
editor_options: 
  markdown: 
    wrap: 72
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = F, warning = F, cache = T)
# requires
# tidyverse - data manipulation 
# VariantAnnotation - to parse VCFs
# reactable - to print tables in HTML

#read config to get strain names
config_mutants <- yaml::read_yaml("/home/andrei/Data/HeteroR/workflow/config_mutants.yaml")
strains <- unname(unlist(config_mutants$parents))
```

Here I will copy and develop further "Overview" sections from `code_sketch.Rmd`

## Variants

We're interested in (1) resistance genes - resistance selection, (2) other genes - media adaptation

### Plots {.tabset .tabset-pills}

#### All variants

```{r}
library(tidyverse)
#library(VariantAnnotation)
#library(GenomicRanges)

path_results <- "/home/andrei/Data/HeteroR/results/mutants/variants/"

vcf_files <- dir(path_results, pattern = "*_filtered.vcf", recursive = T)

var_len <- map_dbl(vcf_files, ~ length(
  VariantAnnotation::readVcf(paste0(path_results, .))))

hist(var_len, breaks = 78, main = "", xlab = "N variants", ylab = "count")
```

Number of strains containing more than 5 variants is `r length(var_len[var_len > 5])`

#### Variants inside genes

```{r}
# where the TSV files are (= genes with variants)
path_results <- "/home/andrei/Data/HeteroR/results/mutants/variants/"

tsv_files <- dir(path_results, pattern = "*_variants.tsv", recursive = T)

genes_w_vars <- map_dfr(paste0(path_results, tsv_files),
                        function(x)
                          read_tsv(
                            x,
                            col_names = T,
                            show_col_types = F,
                            col_types = "ciiccc") %>% 
                          mutate(file = x)) %>%
  mutate(strain = str_extract(file, "DA[0-9]*")) %>% 
  dplyr::select(-file) %>% 
  dplyr::rename("gene_start" = start,
                "gene_end" = stop)

# 38 strains that do not have genes with mutations should also be in the table
# length(tsv_files) - length(unique(genes_w_vars$strain))

# this one I need for the histogram below (it should contain 0 variants)
genes_w_vars_all <-
  strains[!strains %in% unique(genes_w_vars$strain)] %>%
  map_dfr(
    ~ tibble(
      contig = c(NA),
      start = c(NA),
      stop = c(NA),
      strand = c(NA),
      gene_name = c(NA),
      locus_tag = c(NA),
      strain = c(.)
    )
  ) %>%
  bind_rows(genes_w_vars)

genes_w_vars_all %>% 
  group_by(strain, contig) %>% 
  dplyr::count() %>% 
  mutate(n.adj = if_else(is.na(contig), 0, n)) %>% 
  ggplot(aes(n.adj)) +
  geom_histogram(binwidth = 0.5, fill = "seagreen", alpha = 0.7) +
  scale_x_continuous(breaks = c(0, 1, 2, 3, 4, 5, 6, 7)) +
  xlab("N variants") +
  ylab("N strains") +
  ggtitle("Variants inside genes")
```

More than 5 variants:

```{r}
genes_w_vars_all %>% 
  group_by(strain, contig) %>% 
  dplyr::count() %>% 
  filter(n > 5)
```

### Annotated variants

Genes' names/products and variants

```{r}
# function to read a VCF file and add strain name
vcf2tibb <- function(file_path, path_prefix){ 
  # make full path to the file
  full_path <- paste0(path_prefix, file_path)
  # read in vcf
  vcf_df <- as_tibble(MatrixGenerics::rowRanges(VariantAnnotation::readVcf(full_path)))
  # extract strain name: the first 7 characters
  vcf_df$strain <- substr(file_path, 1, 7)
  # convert ALT to char
  vcf_df$ALT.char <- unlist(map(vcf_df$ALT, ~as.character(.)))
  # done
  return(vcf_df)
}

# where the VCF files are
path_vars <- "/home/andrei/Data/HeteroR/results/mutants/variants/"
vcf_files <- dir(path_vars, pattern = "*_filtered.vcf", recursive = T)

# parse the files and join them into one table
var_df <- map_dfr(vcf_files, ~ vcf2tibb(., path_prefix = path_vars)) %>% 
  dplyr::select(-c(ALT, paramRangeID)) %>% 
  dplyr::rename("contig" = seqnames,
                "var_start" = start,
                "var_end" = end,
                "ALT" = ALT.char) %>%
  relocate(ALT, .before = QUAL)
  
# Join variants with the genes they are in
# keep the variants which are between the borders of the genes with variants
# function
filter_vars_in_genes <- function(df_var, df_gene) {
  # df_var -> variants no genes
  # df_gene -> genes no variants
  # check headers
  stopifnot(sum(grepl("var_start", colnames(df_var))) == 1)
  
  stopifnot(sum(grepl("gene_start", colnames(df_gene))) == 1)
  
  # filter variants within genes
  out_df <- map2_dfr(df_gene$gene_start,
                     df_gene$gene_end,
                     ~ filter(df_var, between(var_start, .x, .y) |
                                between(var_end, .x, .y)))
  # return a df with vars inside genes
  return(out_df)
}

# keep only variant of strains that have at least 1 variant 
var_df_filt <- var_df %>% 
  filter(strain %in% unique(genes_w_vars$strain))

var_df_list <- split.data.frame(var_df_filt, f = var_df_filt$strain)

gene_df_list <- split.data.frame(genes_w_vars, f = genes_w_vars$strain)

# apply the function
var_inside_genes_df <-
  map2_dfr(var_df_list, gene_df_list, ~ filter_vars_in_genes(.x, .y)) %>%
  distinct()

# Join `genes with variants` with `variants inside genes`

# split vars inside the genes
var_inside_list <- split.data.frame(var_inside_genes_df, f = var_inside_genes_df$strain)

# apply joining and filtering to each strain-table
vars_inside_final <- map2_dfr(
  gene_df_list,
  var_inside_list,
  ~ left_join(.x, .y, by = c("strain", "contig")) %>% filter(
    between(var_start, gene_start, gene_end) |
      between(var_end, gene_start, gene_end)
  ) %>%
    distinct()
) %>%
  relocate(strain, .before = contig)
```


```{r, eval=FALSE}
# to check that n. rows per strain in both tables is the same
full_join(
  vars_inside_final %>%
    group_by(strain) %>%
    dplyr::count(),
  var_inside_genes_df %>%
    group_by(strain) %>%
    dplyr::count(),
  by = "strain"
) %>%
  mutate(diff = n.x - n.y) %>%
  filter(diff != 0)
```


Strain **DA63362** has some discrepancies between the two tables because two variants were mapped to the last position of the 'gene' `OELAHNJC_03152`, which overlaps `OELAHNJC_03153`.

As the result, the same variants end up on twice more rows than I expected.

```{r, eval=FALSE}
# to check it, run
var_inside_genes_df %>% filter(strain == "DA63362") %>% 
  distinct() %>% 
  arrange(var_start)

# in this one you will see these two variants and 'genes'
vars_inside_final %>% filter(strain == "DA63362") %>% 
  filter(locus_tag == "OELAHNJC_03153" | locus_tag == "OELAHNJC_03152") %>% 
  arrange(var_start)
```

### The final table

```{r}
vars_inside_final %>% 
  dplyr::select(-c(width, strand.y, QUAL, FILTER)) %>% 
  reactable::reactable(
    #style = list(fontFamily = "Work Sans, sans-serif", fontSize = "0.675rem"),
    resizable = TRUE,
    filterable = TRUE,
    minRows = 10,
    sortable = TRUE,
    defaultColDef = reactable::colDef(
      cell = function(value)
        format(value, nsmall = 1),
      align = "center",
      minWidth = 35,
      headerStyle = list(background = "#f7f7f8")
    ),
    columns = list(strain = reactable::colDef(
      sticky = "left",
      # Add a right border style to visually distinguish the sticky column
      style = list(borderRight = "1px solid #eee"),
      headerStyle = list(borderRight = "1px solid #eee")
    ))
    ,
    bordered = TRUE,
    highlight = TRUE
  )
```

Number of strains in the table: `r length(unique(vars_inside_final$strain))`

Number of genes in the table: `r length(unique(vars_inside_final$locus_tag))`

Number of genes with functional (scientific) names: `r length(unique(vars_inside_final$gene_name)) - 1`

Some genes got functional names, some - only systematic names aka "locus tags"

### Functional annotations

```{r}
# this function will be used later
parse_annotations <- function(strains_vec) {
  # this function reads annotation TSV files
  # strains_vec: vector of unique strain names
  map_dfr(
    strains_vec,
    ~ read_tsv(
      paste0(
        "/home/andrei/Data/HeteroR/results/annotations/",
        .,
        "/prokka/",
        .,
        "_genomic.tsv"
      ),
      show_col_types = F
    ) %>% 
      filter(ftype == "CDS", !is.na(gene)) %>% 
      dplyr::select(locus_tag, COG, product)
  )
}

genes_annotation <- parse_annotations(unique(vars_inside_final$strain))

# read the InterProScan results
ips <-
  read_tsv(
    "/home/andrei/Data/HeteroR/results/mutants/variants/InterProScan_results_Galaxy.tsv",
    col_names = F,
    show_col_types = F
  ) %>%
  dplyr::select(X1, X4, X6) %>%
  dplyr::rename("locus_tag" = X1) %>%
  pivot_wider(
    names_from = X4,
    values_from = X6,
    values_fn = dplyr::first
  )

# join together IPS, variants and genes' annotations
vars_inside_final %>% 
  filter(!is.na(gene_name)) %>% 
  left_join(genes_annotation, by="locus_tag") %>% 
  dplyr::select(strain, contig,  locus_tag, gene_name, product, COG, REF, ALT) %>% 
  left_join(ips, by = "locus_tag") %>% 
  dplyr::select(-locus_tag) %>% 
  reactable::reactable(
    #style = list(fontFamily = "Work Sans, sans-serif", fontSize = "0.675rem"),
    resizable = TRUE,
    filterable = TRUE,
    minRows = 10,
    sortable = TRUE,
    defaultColDef = reactable::colDef(
      cell = function(value)
        format(value, nsmall = 1),
      align = "center",
      minWidth = 35,
      headerStyle = list(background = "#f7f7f8")
    ),
    columns = list(strain = reactable::colDef(
      sticky = "left",
      # Add a right border style to visually distinguish the sticky column
      style = list(borderRight = "1px solid #eee"),
      headerStyle = list(borderRight = "1px solid #eee")
    ))
    ,
    bordered = TRUE,
    highlight = TRUE
  )
```


## Amplifications

### Table of known genes inside the amplifications

```{r}
path_results <- "/home/andrei/Data/HeteroR/results/mutants/amplifications/"
file_common_name <- "/amplifications_annotated_filtered.tsv"

read_amp_tsv <- function(tsv, strain_name) {
  out_df <- read_tsv(tsv, show_col_types = F, col_names = T)
  out_df$strain <- strain_name
  return(out_df)
}

amp_df <- map_dfr(strains, ~ read_amp_tsv(paste0(path_results, ., file_common_name), .)) %>% 
  relocate(strain, .before = contig)

genes_annotation <- parse_annotations(unique(amp_df$strain))

left_join(amp_df, genes_annotation, by = "locus_tag") %>%
  dplyr::select(strain, contig, strand, start, stop, gene_name, product, COG) %>%
  filter(!is.na(gene_name)) %>%
  reactable::reactable(
    #style = list(fontFamily = "Work Sans, sans-serif", fontSize = "0.675rem"),
    resizable = TRUE,
    filterable = TRUE,
    minRows = 10,
    sortable = TRUE,
    defaultColDef = reactable::colDef(
      cell = function(value)
        format(value, nsmall = 1),
      align = "center",
      minWidth = 35,
      headerStyle = list(background = "#f7f7f8")
    ),
    columns = list(strain = reactable::colDef(
      sticky = "left",
      # Add a right border style to visually distinguish the sticky column
      style = list(borderRight = "1px solid #eee"),
      headerStyle = list(borderRight = "1px solid #eee")
    ))
    ,
    bordered = TRUE,
    highlight = TRUE
  )
```

Type 'Beta' in the search box to find beta-lactamases

How many strains have amplifications: `r length(unique(amp_df$strain))`

Number of functional gene names inside the amplifications: `r length(unique(amp_df$gene_name)) - 1`

Number of systematic gene names inside the amplifications: `r length(unique(amp_df$locus_tag))`

### Plots {.tabset .tabset-pills}

#### Histogram

```{r}
amp_df %>% 
  group_by(strain) %>% 
  dplyr::count() %>% 
  ggplot(aes(n)) + 
  geom_histogram(fill="steelblue", alpha = 0.8, binwidth = 4) +
  scale_y_continuous(breaks = c(0, 2, 4, 6, 8)) +
  xlab("N genes") +
  ylab("N strains") +
  ggtitle("Genes inside amplifications (per strain)")
```

#### Barplot

```{r, fig.width=10}
amp_df %>% 
  group_by(strain) %>% 
  dplyr::count() %>% 
  ggplot(aes(reorder(strain, -n), n)) +
  geom_col(fill = "steelblue") +
  theme(axis.text.x = element_text(size=7, angle=45, hjust = 1)) +
  xlab("") +
  ylab("N genes with amplifications") +
  ggtitle("Number of genes in amplifications")
```

## Plasmid copy number change

### Overview of contigs' lengths

```{r}
# code from code_sketch.Rmd
path_results <- "/home/andrei/Data/HeteroR/results/mutants/copy_number/"

file_common_name <- "/relative_coverage"

# parents
rel_cov_df_parent <-
  map_dfr(strains,
          ~ read_tsv(
            paste0(path_results,
                   ., file_common_name, "_parent.tsv"),
            show_col_types = F
          ))

# mutants
rel_cov_df_mutant <-
  map_dfr(strains,
          ~ read_tsv(
            paste0(path_results,
                   ., file_common_name, "_mutant.tsv"),
            show_col_types = F
          ))

# altogether, minus chromosomes
# this table is required for the plots below
rel_cov_final <- bind_rows(rel_cov_df_mutant, rel_cov_df_parent) %>%
  filter(genome.part != "chrom") %>%
  mutate(contig.name.2 = str_extract(contig.name, "_[0-9]"),
         contig.name.2 = sub(pattern = "_", replacement = "", x = contig.name.2, fixed = T)) %>%
  relocate(strain, .before = contig.name)

# rel_cov_final %>%
#   # keep this select() here!
#   dplyr::select(-c(genome.part, contig.name.2)) %>% 
#   reactable::reactable(
#     filterable = TRUE,
#     minRows = 10,
#     sortable = TRUE,
#     defaultColDef = reactable::colDef(
#       cell = function(value)
#         format(value, nsmall = 1),
#       align = "center",
#       minWidth = 70,
#       headerStyle = list(background = "#f7f7f8")
#     ),
#     bordered = TRUE,
#     highlight = TRUE
#   )
```



```{r}
rel_cov_final %>% 
  ggplot(aes(contig.len)) +
  geom_histogram(bins = 100) +
  geom_rug() +
  geom_vline(xintercept = 100000, lty = 3) +
  ggtitle("Length distribution of all contigs")
```

Some contigs are too long for plasmids. Vertical dotted line is at 100 kb.


### Relative depth comparison plots {.tabset .tabset-pills}

Contigs longer than 100 kb were filtered out

#### Plot 1

```{r, fig.height=6, fig.width=8}
plot_bars <- function(df, parent_name) {
  pl <- df %>%
    dplyr::filter(strain == parent_name) %>% 
    ggplot(aes(contig.name.2, round(rel.depth, 1))) +
    geom_col(aes(fill = label), position = "dodge") +
    ggtitle(parent_name) +
    xlab("plasmid") +
    ylab("relative depth") +
    theme(legend.position = "none") 
  
  return(pl)
}

# filtering here
plot_list <- map(strains, ~ plot_bars(rel_cov_final %>% filter(contig.len < 100000), .))

ggpubr::ggarrange(plotlist = plot_list[c(1:6)], nrow = 2, ncol = 3, common.legend = T)
```

#### Plot 2

```{r, fig.height=6, fig.width=8}
ggpubr::ggarrange(plotlist = plot_list[c(7:12)], nrow = 2, ncol = 3, common.legend = T)
```

#### Plot 3

```{r, fig.height=6, fig.width=8}
ggpubr::ggarrange(plotlist = plot_list[c(13:18)], nrow = 2, ncol = 3, common.legend = T)
```

#### Plot 4

```{r, fig.height=6, fig.width=8}
ggpubr::ggarrange(plotlist = plot_list[c(19:24)], nrow = 2, ncol = 3, common.legend = T)
```


#### Plot 5

```{r, fig.height=6, fig.width=8}
ggpubr::ggarrange(plotlist = plot_list[c(25:30)], nrow = 2, ncol = 3, common.legend = T)
```


#### Plot 6

```{r, fig.height=6, fig.width=8}
ggpubr::ggarrange(plotlist = plot_list[c(31:36)], nrow = 2, ncol = 3, common.legend = T)
```


#### Plot 7

```{r, fig.height=6, fig.width=8}
ggpubr::ggarrange(plotlist = plot_list[c(37:42)], nrow = 2, ncol = 3, common.legend = T)
```

#### Plot 8

```{r, fig.height=6, fig.width=8}
ggpubr::ggarrange(plotlist = plot_list[c(43:48)], nrow = 2, ncol = 3, common.legend = T)
```

#### Plot 9

```{r, fig.height=6, fig.width=8}
ggpubr::ggarrange(plotlist = plot_list[c(49:54)], nrow = 2, ncol = 3, common.legend = T)
```

#### Plot 10

```{r, fig.height=6, fig.width=8}
ggpubr::ggarrange(plotlist = plot_list[c(55:60)], nrow = 2, ncol = 3, common.legend = T)
```

#### Plot 11

```{r, fig.height=6, fig.width=8}
ggpubr::ggarrange(plotlist = plot_list[c(61:66)], nrow = 2, ncol = 3, common.legend = T)
```


#### Plot 12

```{r, fig.height=6, fig.width=8}
ggpubr::ggarrange(plotlist = plot_list[c(67:72)], nrow = 2, ncol = 3, common.legend = T)
```


### Table of all genes

```{r}
# these annotations do not have contig name
genes_annotation <- parse_annotations(unique(rel_cov_final$strain))

# read GFF additionally to get contig names
read_gff <- function(gff_filename) {
    as_tibble(read.delim(gff_filename, header = F, comment.char = "#")) %>%
    dplyr::select(V1, V3, V4, V5, V7, V9) %>%
    filter(V3 == "gene")
}

gff_table <- map_dfr(
  unique(rel_cov_final$strain), 
  ~ read_gff(paste0("/home/andrei/Data/HeteroR/results/annotations/",
                    .,
                    "/prokka/", ., "_genomic.gff")) %>% 
    # keep only needed columns
    dplyr::select(V1, V9)) %>% 
    # extract locus tags from V9
    mutate(locus_tag = str_extract(V9, "[A-Z]*_[0-9]*")) %>% 
    # remove V9
    dplyr::select(-V9) %>% 
    dplyr::rename("contig.name" = V1)

# add the contig names to the annotations
genes_annotation <- left_join(genes_annotation, gff_table, by = "locus_tag")

# join genes to contigs with their relative coverage
left_join(rel_cov_final, genes_annotation, by = "contig.name") %>%
  # filter(contig.len < 100000)
  # filtering?
  dplyr::select(strain,
         contig.name,
         contig.len,
         rel.depth,
         locus_tag,
         label,
         product) %>%
  pivot_wider(names_from = label, values_from = rel.depth) %>%
  dplyr::rename("rel.depth.mutant" = mutant,
         "rel.depth.parent" = parent) %>%
  relocate(rel.depth.mutant, rel.depth.parent, .after = contig.len) %>%
  mutate(
    rel.depth.mutant = round(rel.depth.mutant, 1),
    rel.depth.parent = round(rel.depth.parent, 1),
    CN = if_else(rel.depth.mutant > rel.depth.parent, "increased", "decreased")
  ) %>%
  reactable::reactable(
    #style = list(fontFamily = "Work Sans, sans-serif", fontSize = "0.675rem"),
    resizable = TRUE,
    filterable = TRUE,
    minRows = 10,
    sortable = TRUE,
    defaultColDef = reactable::colDef(
      cell = function(value)
        format(value, nsmall = 1),
      align = "center",
      minWidth = 35,
      headerStyle = list(background = "#f7f7f8")
    ),
    columns = list(strain = reactable::colDef(
      sticky = "left",
      # Add a right border style to visually distinguish the sticky column
      style = list(borderRight = "1px solid #eee"),
      headerStyle = list(borderRight = "1px solid #eee")
    ))
    ,
    bordered = TRUE,
    highlight = TRUE
  )
```


### Table of genes on contings shorter than 100 kb

```{r}
left_join(rel_cov_final, genes_annotation, by = "contig.name") %>%
  # filter(contig.len < 100000)
  # filtering?
  dplyr::select(strain,
         contig.name,
         contig.len,
         rel.depth,
         locus_tag,
         label,
         product) %>%
  filter(contig.len < 100000) %>%
  pivot_wider(names_from = label, values_from = rel.depth) %>%
  dplyr::rename("rel.depth.mutant" = mutant,
         "rel.depth.parent" = parent) %>%
  relocate(rel.depth.mutant, rel.depth.parent, .after = contig.len) %>%
  mutate(
    rel.depth.mutant = round(rel.depth.mutant, 1),
    rel.depth.parent = round(rel.depth.parent, 1),
    CN = if_else(rel.depth.mutant > rel.depth.parent, "increased", "decreased")
  ) %>%
  reactable::reactable(
    #style = list(fontFamily = "Work Sans, sans-serif", fontSize = "0.675rem"),
    resizable = TRUE,
    filterable = TRUE,
    minRows = 10,
    sortable = TRUE,
    defaultColDef = reactable::colDef(
      cell = function(value)
        format(value, nsmall = 1),
      align = "center",
      minWidth = 35,
      headerStyle = list(background = "#f7f7f8")
    ),
    columns = list(strain = reactable::colDef(
      sticky = "left",
      # Add a right border style to visually distinguish the sticky column
      style = list(borderRight = "1px solid #eee"),
      headerStyle = list(borderRight = "1px solid #eee")
    ))
    ,
    bordered = TRUE,
    highlight = TRUE
  )
```

## IS elements 

### Overview of IS found in parental strains

```{r}
path_results <- "/home/andrei/Data/HeteroR/results/isescan/"
file_common_name <- "/regions/regions_joined_final.fasta.tsv"

is_df <-
  map_dfr(strains, ~ tryCatch({ read_tsv(paste0(path_results, ., file_common_name), show_col_types = F)}, error = function(e) {
      message(paste("No file exist for ", ., sep = " "))
      return(NULL)}))
# No file exist for  DA63688

# read family lengths from the yaml file
ise_lengths <- yaml::read_yaml("is_lengths.yaml")
# find IS length according to DB
ise_len_df <-
  tibble(family = names(ise_lengths),
         db.length = as.numeric(sapply(c(1:20), function(x)
           ise_lengths[[x]])))

# add db.lengths to the main table
is_df <- left_join(is_df, ise_len_df, by="family")

```


### Histograms {.tabset .tabset-pills}

Vertical line represents average length of a particular element according ISfinder database

```{r}
is_lst <- split.data.frame(is_df, f = is_df$family)

plot_length_hist <- function(df) {
  # df should contain isLen, orfLen, family, db.length
  out_plot <- df %>%
    ggplot(aes(obs.length)) +
    geom_histogram(bins = 50, aes(fill = label)) +
    geom_rug(aes(color = label)) +
    geom_vline(aes(xintercept = db.length)) +
    ggtitle(df$family[1])
  
  return(out_plot)
}

# filter, reshape and add db.length column
reshape_is_df <- function(is_df) {
  out_df <- is_df %>% 
    dplyr::select(isLen, orfLen, family, db.length) %>% 
    pivot_longer(cols = c(1, 2), names_to = "label", values_to = "obs.length")
  #hist_plot <- plot_length_hist(out_df)
  return(out_df)
}

# reshape every table and make a list of histograms
is_hist_list <- map(is_lst, ~ plot_length_hist(reshape_is_df(.)))
```

#### IS1 & IS110

```{r}
ggpubr::ggarrange(plotlist = is_hist_list[c(1,2)], ncol = 1, nrow = 2, common.legend = T, legend = "right")
```

#### IS1182 & IS1380

```{r}
ggpubr::ggarrange(plotlist = is_hist_list[c(3,4)], ncol = 1, nrow = 2, common.legend = T, legend = "right")
```

#### IS200/IS605 & IS21

```{r}
ggpubr::ggarrange(plotlist = is_hist_list[c(5, 6)], ncol = 1, nrow = 2, common.legend = T, legend = "right")
```

#### IS256 & IS3

```{r}
ggpubr::ggarrange(plotlist = is_hist_list[c(7, 8)], ncol = 1, nrow = 2, common.legend = T, legend = "right")
```

#### IS30 & IS4

```{r}
ggpubr::ggarrange(plotlist = is_hist_list[c(9,10)], ncol = 1, nrow = 2, common.legend = T, legend = "right")
```

#### IS481 & IS5

```{r}
ggpubr::ggarrange(plotlist = is_hist_list[c(11,12)], ncol = 1, nrow = 2, common.legend = T, legend = "right")
```

#### IS6 & IS630

```{r}
ggpubr::ggarrange(plotlist = is_hist_list[c(13,14)], ncol = 1, nrow = 2, common.legend = T, legend = "right")
```

#### IS66 & IS91

```{r}
ggpubr::ggarrange(plotlist = is_hist_list[c(15,16)], ncol = 1, nrow = 2, common.legend = T, legend = "right")
```

#### ISAS1 & ISL3

```{r}
ggpubr::ggarrange(plotlist = is_hist_list[c(17,18)], ncol = 1, nrow = 2, common.legend = T, legend = "right")
```

#### ISNCY & 'new'

```{r}
ggpubr::ggarrange(plotlist = is_hist_list[c(19,20)], ncol = 1, nrow = 2, common.legend = T, legend = "right")
```

### New insertions in the mutants

```{r}
results_path <- "/home/andrei/Data/HeteroR/results/mutants/ismapper/new_insertions"
results_files <- dir(path = results_path, pattern = "*table.txt", recursive = T, full.names = T)

best_is_representatives <- read_tsv("/home/andrei/Data/HeteroR/results/mutants/ismapper/best_IS_representatives.tsv")

# read the provided table and add IS name
ismapper_table <- function(table_file, family_id, strain_name) {
  # table_file: a file with ISmapper results
  # family_id: a data frame with two columns 'seqID' and 'family' from 'best_is_representatives'
  out_df <- read_tsv(table_file, show_col_types = F, col_types = "ccdddccccccdcccdc") 
  # find family name
  seqid <- str_extract(table_file, "[A-Z]*_[0-9]*_gene")
  fam_name <- dplyr::filter(family_id, seqID == seqid) %>% pull(family)
  out_df$family <- fam_name
  out_df$strain <- strain_name
  return(out_df)
}

new_inserts <- map_dfr(results_files,
         ~ ismapper_table(
           table_file = .,
           family_id = dplyr::select(best_is_representatives, seqID, family),
           strain_name = str_extract(., "DA[0-9]*")
         ))


# For the three problematic strains another function is needed
# Because they have different patterns in the output file names
results_path2 <- "/home/andrei/Data/HeteroR/results/mutants/ismapper/separate_test_results"
results_files2 <- dir(path = results_path2, pattern = "*table.txt", recursive = T, full.names = T)

# IS family is present in the file name itself
ismapper_table2 <- function(table_file, strain_name) {
  # table_file: a file with ISmapper results
  out_df <- read_tsv(table_file, show_col_types = F, col_types = "ccdddccccccdcccdc") 
  out_df$family <- str_extract(table_file, "IS[0-9,A-Z]*")
  out_df$strain <- strain_name
  return(out_df)
}

new_inserts2 <-
  map_dfr(results_files2,
          ~ ismapper_table2(table_file = ., 
                            strain_name = str_extract(., "DA[0-9]*")))
```

```{r}
# join and print
bind_rows(new_inserts, new_inserts2) %>%
  filter(
    left_description != "hypothetical protein",
    right_description != "hypothetical protein",
    call == "novel*"
  ) %>%
  dplyr::select(-c(gap, percent_ID, percent_cov, call, contains("strand"))) %>%
  mutate(
    region = str_extract(region, "[0-9]"),
    x = round(x, 0),
    y = round(y, 0)
  ) %>%
  unite("pos", x:y, sep = " - ", remove = T) %>% 
  relocate(c(strain, family), .before = region) %>%
  reactable::reactable(
    #style = list(fontFamily = "Work Sans, sans-serif", fontSize = "0.675rem"),
    resizable = TRUE,
    filterable = TRUE,
    minRows = 10,
    sortable = TRUE,
    defaultColDef = reactable::colDef(
      cell = function(value)
        format(value, nsmall = 1),
      align = "center",
      minWidth = 35,
      headerStyle = list(background = "#f7f7f8")
    ),
    columns = list(strain = reactable::colDef(
      sticky = "left",
      # Add a right border style to visually distinguish the sticky column
      style = list(borderRight = "1px solid #eee"),
      headerStyle = list(borderRight = "1px solid #eee")
    ))
    ,
    bordered = TRUE,
    highlight = TRUE
  )
```


#### Plots

How many new insertions in proximity to known genes per strain?

```{r, fig.width=10}
bind_rows(new_inserts, new_inserts2) %>% 
  filter(left_description != "hypothetical protein",
         right_description != "hypothetical protein",
         call == "novel*") %>% 
  dplyr::select(strain, region) %>% 
  distinct() %>% 
  group_by(strain) %>% 
  dplyr::count() %>% 
  ggplot(aes(reorder(strain, -n), n)) +
  geom_col(fill = "steelblue") +
  theme(axis.text.x = element_text(size = 6, angle = 45, hjust = 1)) +
  xlab("") +
  ylab("new instertions")
```

How many genes next to each IS family?

```{r}
bind_rows(new_inserts, new_inserts2) %>% 
  filter(left_description != "hypothetical protein",
         right_description != "hypothetical protein",
         call == "novel*") %>% 
  group_by(family) %>% 
  dplyr::count() %>% 
  ggplot(aes(reorder(family, -n), n)) +
  geom_col(fill = "steelblue") +
  theme(axis.text.x = element_text(size = 12, angle = 45, hjust = 1)) +
  xlab("")
```

How many gene interruptions?

```{r, fig.width=10}
bind_rows(new_inserts, new_inserts2) %>% 
  filter(left_description != "hypothetical protein",
         right_description != "hypothetical protein",
         gene_interruption == "True",
         call == "novel*") %>% 
  dplyr::select(strain, region) %>% 
  distinct() %>% 
  group_by(strain) %>% 
  dplyr::count() %>% 
  ggplot(aes(reorder(strain, -n), n)) +
  geom_col(fill = "steelblue") +
  theme(axis.text.x = element_text(size = 6, angle = 45, hjust = 1)) +
  xlab("")
```

