---
title: "variant calling tests"
author: "AG"
date: "2023-11-21"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Variants

## Mapping

```{bash, eval=F}
mamba env create --file var_calling.yaml
```

### Take fasta records from gff annotation file

because they contain correct contig names!

```{bash make.fasta, eval=F}
GFF="/home/andrei/Data/HeteroR/results/annotations/DA63342/prokka/DA63342_genomic.gff"

# leave fasta, drop the rest
sed -n '/##FASTA/,${p}' $GFF | sed '1d' > assembly_new_names.fasta 
```

```{bash mapping, eval=F}
READS1="/home/andrei/Data/HeteroR/results/data_filtered/DA63342/Illumina/mutants/DA63342m_1.fq.gz"
READS2="/home/andrei/Data/HeteroR/results/data_filtered/DA63342/Illumina/mutants/DA63342m_2.fq.gz"
REF_FASTA="assembly_new_names.fasta"
# make index first!
# mapping
cd results/variants
conda activate varcalling-env
bowtie2-build $REF_FASTA index
bowtie2 -p 10 -x index -1 $READS1 -2 $READS2 -S aligned.sam && samtools sort -@ 10 aligned.sam > aligned.bam
```

## Variant calling & filtering

```{bash var.calling, eval=F}
bcftools mpileup --threads 10 --max-depth 800 -f $REF_FASTA -Ou aligned.bam | bcftools call --threads 10 --ploidy 1 -mv -Ob --prior 5.0e-10 -o variants.bcf
bcftools filter -g3 -i 'QUAL>30 && DP>20' -Ob variants.bcf > variants_filtered.bcf
# to take a look
bcftools index variants_filtered.bcf
bcftools query --format "%POS\t%REF\t%ALT\n" variants_filtered.bcf
```

## Variant annotation

```{bash var.annotation, eval=F}
GFF="/home/andrei/Data/HeteroR/results/annotations/DA63342/prokka/DA63342_genomic.gff"

bcftools view variants_filtered.bcf > variants_filtered.vcf

# drop the genome fasta in the gff file
sed '/##FASTA/,$d' $GFF > DA63342_genomic_clean.gff 

# vcf file must have the same contig names as the gff file
bedtools annotate -i DA63342_genomic_clean.gff -files variants_filtered.vcf > gff_with_annot.tsv
# "Default behavior is to report the fraction of -i covered by each file"
```

```{r read.gff, eval=F}
library(dplyr)

read_gff <- function(gff_filename) {
  # read and process
  df <-
    as_tibble(read.delim(gff_filename, header = F, comment.char = "#")) %>%
    select(V1, V3, V4, V5, V7, V9, V10) %>%
    filter(V3 == "gene")
  return(df)
}

gff_annotated <- read_gff("/mnt/data/andrei/Data/HeteroR/results/variants/DA63342/DA63342_annotated_variants.gff") %>% filter(V10 > 0)
gff_annotated
```


## Overview of the variants

let's check how many and what kind of variants we've got in our 78 strains

First 10 bases in all reads were trimmed - they had very skewed nucleotide content.

```{r, message=FALSE}
library(tidyverse)
library(VariantAnnotation)
library(GenomicRanges)

path_results <- "/home/andrei/Data/HeteroR/results/mutants/variants/"

vcf_files <- dir(path_results, pattern = "*.vcf", recursive = T)


var_len <- map_dbl(vcf_files, ~ length(
  readVcf(paste0(path_results, .))))

hist(var_len, breaks = 78)
```

```{r}
length(var_len)
```


Number of variants

```{r}
var_len
```

```{r}
summary(var_len)
```


Turn VCF to tibbles

```{r}
vcf2tibb <- function(file_path, path_prefix){
  # make full path to the file
  full_path <- paste0(path_prefix, file_path)
  # read in vcf
  vcf_df <- as_tibble(rowRanges(readVcf(full_path)))
  # extract strain name: the first 7 characters
  vcf_df$strain <- substr(file_path, 1, 7)
  return(vcf_df)
}

var_df <- map_dfr(vcf_files, ~ vcf2tibb(., path_prefix = path_results))

var_df
```

Hist of rows per strain should be the same as the one above

```{r}
var_df %>% 
  group_by(strain) %>% 
  summarise(n = n()) %>% 
  #count() %>% 
  ggplot(aes(n)) +
  geom_histogram(bins = 78) +
  geom_rug() +
  xlab("N variants") +
  ylab("N strains")
```

```{r}
var_df %>% 
  group_by(strain) %>% 
  summarise(n = n()) %>% 
  pull(n) %>% 
  summary()
```


### Variant counts per strain

```{r, fig.width=6, fig.height=3}
var_df %>% 
  group_by(strain) %>% 
  summarise(n = n()) %>% 
  #count() %>% 
  ggplot(aes(strain, n)) +
  geom_col() + 
  theme(axis.text.x = element_text(size = 10, angle = 45, hjust = 1)) +
  geom_hline(yintercept = 3)
```

Ordered by n

```{r}
var_df %>% 
  group_by(strain) %>% 
  dplyr::count() %>% 
  arrange(-n)
```

The top two have currently problems with assembling

###  Strains with too many variants

```{r}
var_df %>% 
  filter(strain == "DA63790") %>% 
  group_by(seqnames) %>% 
  dplyr::count()
```


```{r}
var_df %>% 
  filter(strain == "DA63694") %>% 
  group_by(seqnames) %>% 
  dplyr::count()
```

### DA63346

The strain that is not in the HR testing results...

Before I discovered this, it had approx 7 variants

```{r}
var_df %>% 
  filter(strain == "DA63346")
```

Everything with more than 10 variants

```{r}
var_df %>% 
  group_by(strain, seqnames) %>% 
  dplyr::count() %>% 
  filter(n > 10) %>% 
  arrange(-n)
```

Re-run the pipelines on these 12 strains

```{r, eval=F}
var_df %>% 
  group_by(strain, seqnames) %>% 
  dplyr::count() %>% 
  filter(n > 10) %>%
  pull(strain) %>% 
  write.csv(file = "/home/andrei/Data/HeteroR/resources/strain_lists/mutants_10vars_rerun.txt", quote = F, row.names = F)
```

Copy the corresponding original files using temporary script 'copy_mutants.py'


#### What I did with DA63126 (and others)

1. fastqc once again: DA63126 has too big GC content deviations. Contamination?
2. fastq_screen
3. Filtering of E.coli mapping only reads
4. get VCF again (n = 2400, before n = 4000)

Code:

```{bash, eval=F}
cd results/data_filtered/DA63126/Illumina/mutants
conda activate fastqscreen-env
# tag reads E.coli reads and keep them, toss the rest
fastq_screen --tag --filter 00000001000000 --conf FastQ_Screen_Genomes/fastq_screen.conf DA63126m_1.fq.gz --outdir filter_ecoli
# genome order: hum, mus, rat, dro, worm, yeast, arabido, ecoli, rrna, mito, phix, lamb, vec, adap

# now equalize number of reads in each fastq file
seqkit pair --id-regexp '^(\S+)\/[12]'  -1 filter_ecoli/DA63126m_1.fq.gz -2 filter_ecoli/DA63126m_2.fq.gz

# move the balanced and filtered reads to . and run the pipeline again
# run var_calling.Rmd -> number of variants in DA63126 is 2400, still too much
```


Plots from DA63126 fastq screen:

[screening results reads 1](~/Data/HeteroR/results/data_filtered/DA63126/Illumina/mutants/filter_ecoli/DA63126m_1_screen.png)

[screening results reads 2](~/Data/HeteroR/results/data_filtered/DA63126/Illumina/mutants/filter_ecoli/DA63126m_2_screen.png)

Screening of the other suspicious strains and one of the *clean* strains (DA62954) gives the same looking plots.

From DA63784 fastq screen (200 variants):

[screening results reads 1](~/Data/HeteroR/results/data_filtered/DA63784/Illumina/mutants/filter_ecoli/DA63784m_1_screen.png)


From DA62954 fastq screen (1 variant)

[screening results reads 1](~/Data/HeteroR/results/data_filtered/DA62954/Illumina/mutants/filter_ecoli/DA62954m_1_screen.png)



# Amplifications

## Depth and amplifications

```{r, fig.width=14}
library(data.table)

# depth.txt is made by samtools depth/coverage (?)
depth <- fread("/home/andrei/Data/HeteroR/results/variants/DA63150/depth.tsv.gz", sep = "\t")
#depth <- fread("/home/andrei/Data/HeteroR/results/variants/DA63342/depth.tsv.gz", sep = "\t")
# which one is chromosome?
depth %>% group_by(V1) %>% count()
depth_chosen <- depth[V1 == "gnl|UU|CMIAEBAJ_1",]
#depth_chosen <- depth[V1 == "gnl|UU|JBJGLNML_2", ]
depth_chosen
```


## coverage in a window of certain size

calculate z-score

+ 2SD is your candidate

>The absolute value of z represents the distance between that raw score x and the population mean in units of the standard deviation. z is negative when the raw score is below the mean, positive when above.

```{r, fig.width=14, fig.height=4}
#library(dplyr)
library(ggplot2)

window_size <- 1000
mean_coverage <- mean(depth_chosen$V3)

depth_chosen %>%
  mutate(window = 1 + (row_number() - 1) %/% window_size) %>%
  group_by(window) %>%
  summarise(coverage = mean(V3)) %>%
  mutate(z = (coverage - mean(coverage)) / sd(coverage)) %>%
  ggplot(aes(window, z)) +
  geom_point(size = 0.2) +
  geom_hline(yintercept = 2, color = "red")
```


```{r}
depth_per_window <-
        depth_chosen %>%
  mutate(window = 1 + (row_number() - 1) %/% window_size) %>%
  group_by(window) %>%
  summarise(coverage = mean(V3)) %>%
  mutate(z = (coverage - mean(coverage)) / sd(coverage))

depth_per_window %>%
  filter(z >= 3) %>%
        arrange(window)
```


# Plasmid CN

via change in coverage compared to the parental strains

1. You have depth.txt files with short reads coverage for each parent -> summarise

2. You have coverage files for mutants (from the 'amplifications' step) -> get coverage; summarise

3. Join and compare

## Parents

```{r, warning=FALSE}
library(tidyverse)

config_mutants <- yaml::read_yaml("/home/andrei/Data/HeteroR/workflow/config_mutants.yaml")

parents <- unlist(config_mutants$parents)

dir_path <- "/home/andrei/Data/HeteroR/results/mutants/copy_number/"

test_depth <- paste0(dir_path, parents[53], "/parent_depth.tsv.gz")

test_depth_df <- read_tsv(test_depth, col_names = F)

head(test_depth_df)
```

Make a table with element lengths, absolute, average and relative (to chromosome) coverage 

```{r}
cov_df_parent <- 
  test_depth_df %>% 
  group_by(X1) %>% 
  summarise(length = n(), 
            abs.cov = sum(X3)) %>% 
  filter(length > 1000) %>% 
  mutate(avg.cov = round(abs.cov/length, 3))

# relative coverage
chrom_cov_parent <- cov_df_parent %>% 
  filter(length == max(length)) %>% 
  pull(avg.cov)

cov_df_parent$rel.cov <- cov_df_parent$avg.cov/chrom_cov_parent

cov_df_parent <-
  cov_df_parent %>%
  arrange(-length)

cov_df_parent
```

## Mutants
 
You have depth file in `amplifications`

```{r}
# cd results/variants/DA62886
# conda activate varcalling-env
# samtools index mutant_mapped.bam 
# samtools depth -a mutant_mapped.bam > mutant_depth.tsv 
mut_depth_df <- read_tsv("/home/andrei/Data/HeteroR/results/mutants/amplifications/DA63752/mutant_depth.tsv.gz", col_names = F, show_col_types = F)

mut_depth_df %>% 
  group_by(X1) %>% 
  dplyr::count()
```

```{r}
cov_df_mutant <- 
  mut_depth_df %>% 
  group_by(X1) %>% 
  summarise(length = n(), 
            abs.cov = sum(X3)) %>% 
  filter(length > 1000) %>% 
  mutate(avg.cov = round(abs.cov/length, 3))

# relative coverage
chrom_cov_mutant <- cov_df_mutant %>% 
  filter(length == max(length)) %>% 
  pull(avg.cov)

cov_df_mutant$rel.cov <- cov_df_mutant$avg.cov/chrom_cov_mutant

cov_df_mutant <- cov_df_mutant %>% arrange(-length)

cov_df_mutant
```

### Join & Compare

```{r}
bind_cols(cov_df_parent, cov_df_mutant) %>% 
  select(contains("X1"), contains("rel.cov"), contains("length"))
```

### Function

This function is used in a script in the pipeline

```{r}
parse_depth <- function(file_path, min_contig_length=1000, strain, label) {
  # param file_path: depth file path
  # min_contig_length: contigs to filter out
  # param strain: strain name
  # param label: mutant or parent
  depth_df <- read_tsv(file_path, col_names = F, show_col_types = F)
  
  # check correct column names
  stopifnot(exprs = {"X1" %in% names(depth_df) & "X3" %in% names(depth_df)})
  # check if the table is empty
  stopifnot(nrow(depth_df) > 0)
  
  # calculate length and avg. coverage of each contig
  cov_df <-
    depth_df %>%
    group_by(X1) %>%
    summarise(length = n(),
              coverage = sum(X3)) %>%
    filter(length > min_contig_length) %>%
    mutate(avg.cov = round(coverage / length, 3)) %>% 
    arrange(-length)
  
  # relative coverage
  chrom_cov <- cov_df %>%
    filter(length == max(length)) %>%
    pull(avg.cov)
  
  cov_df$rel.cov <- cov_df$avg.cov / chrom_cov
  cov_df$strain <- strain
  cov_df$label <- label
  chrom_plasm_col <- c("chrom", rep("plasmid", nrow(cov_df) - 1))
  cov_df$genome.part <- chrom_plasm_col
  
  return(cov_df)
}

mut <- parse_depth("/home/andrei/Data/HeteroR/results/mutants/amplifications/DA62916/mutant_depth.tsv.gz", strain = "DA62916", label = "mutant")
parent <- parse_depth("/home/andrei/Data/HeteroR/results/mutants/copy_number/DA62916/parent_depth.tsv.gz", strain = "DA62916", label = "parent")

bind_rows(parent, mut) %>% 
  filter(genome.part != "chrom") %>% 
  ggplot(aes(strain, rel.cov)) +
  geom_col(aes(fill=label), position = "dodge")
```

Or look at them as a fold change

```{r}
full_join(parent, mut, by="X1") %>% 
  mutate(rel.cov.fold.change = round(rel.cov.y / rel.cov.x, 2)) %>% 
  select(X1, genome.part.x, rel.cov.fold.change, strain.x) %>% 
  filter(genome.part.x != "chrom") %>% 
  ggplot(aes(X1, rel.cov.fold.change)) +
  geom_col() +
  facet_grid(rows=vars(strain.x)) +
  xlab("") +
  ylab("relative coverage fold change")


```


## Overview of results

After applying the script calculating relative coverage for each mutant strain

```{r, message=FALSE}
library(tidyverse)

config_mutants <- yaml::read_yaml("/home/andrei/Data/HeteroR/workflow/config_mutants.yaml")
parents <- unname(unlist(config_mutants$parents))


rel_cov_df_parent <-
  purrr::map_dfr(parents,
                 ~ readr::read_tsv(paste0(
                       "/home/andrei/Data/HeteroR/results/mutants/copy_number/",
                       ., "/relative_coverage_parent.tsv"), show_col_types = F
                   ))

rel_cov_df_parent
```

```{r}
rel_cov_df_mutant <-
  purrr::map_dfr(parents,
                 ~ readr::read_tsv(paste0(
                       "/home/andrei/Data/HeteroR/results/mutants/copy_number/",
                       ., "/relative_coverage_mutant.tsv"), show_col_types = F
                   ))

rel_cov_df_mutant
```

### Depth files have unreported positions

Why do they have different number of rows?

```{r}
full_join(
  rel_cov_df_parent %>%
    group_by(strain, X1) %>%
    count(),
  rel_cov_df_mutant %>%
    group_by(strain, X1) %>%
    count(),
  by = c("strain","X1")
) %>%
  filter(is.na(n.y) | is.na(n.x))
```

```{r}
rel_cov_df_mutant %>% 
  filter(strain == "DA63038")
```

```{r}
rel_cov_df_parent %>% 
  filter(strain == "DA63038")
```

Why so different? Reference file was the same.

Possible explanation: coverage of the mutant has "gaps" and such regions are not reported, therefore the length is calculated incorrectly or is NA (i.e. gnl|UU|DLCOMDEO_3 is not present in the table - not covered at all)

If depth file has non-covered region, how are they reported? as 0-s or as nothing? (not reported) and then there are gaps in X1 column of the depth file?

```{r}
mut <- data.table::fread("/home/andrei/Data/HeteroR/results/mutants/amplifications/DA63038/mutant_depth.tsv.gz", sep = "\t")
mut[, ':=' (strain="DA63038", label="mutant")]

parent <- parse_depth("/home/andrei/Data/HeteroR/results/mutants/copy_number/DA63038/parent_depth.tsv.gz", strain = "DA63038", label = "parent")

```


```{r}
# any non reported positions?
mut %>% 
  filter(V1 == "gnl|UU|DLCOMDEO_1") %>% 
  mutate(V2.lag = lag(V2),
         V2.diff = V2 - V2.lag) 
```

```{r}
mut %>% 
  filter(V1 == "gnl|UU|DLCOMDEO_1") %>% 
  mutate(V2.lag = lag(V2),
         V2.diff = V2 - V2.lag) %>% 
  filter(V2.diff > 1)
```

So, some positions are not reported because they have 0 coverage.

```{r}
mut %>% 
  filter(V1 == "gnl|UU|DLCOMDEO_1") %>% 
  mutate(V2.lag = lag(V2),
         V2.diff = V2 - V2.lag) %>% 
  filter(V2 <= 147407, V2 >= 147394)
```

I should account for this when calculating length and average coverage...

### Reading length directly from reference

Calculate length

```{r}
library(Biostrings)

file_path <- "/home/andrei/Data/HeteroR/results/mutants/variants/DA63038/reference.fasta" 

# Read the file
ref_gen <- readDNAStringSet(file_path)
contig_lengths <- width(ref_gen)

contig_len_df <- tibble("contig.name" = names(ref_gen), 
                       "contig.len" = contig_lengths)
contig_len_df
```

Calculate coverage

```{r}
mut_depth_df <- read_tsv("/home/andrei/Data/HeteroR/results/mutants/amplifications/DA63038/mutant_depth.tsv.gz", col_names = F, show_col_types = F)

mut_depth_df %>% 
  group_by(X1) %>% 
  summarise(tot.depth = sum(X3)) %>% 
  rename("contig.name" = X1) %>% 
  left_join(contig_len_df, by = "contig.name") %>% 
  mutate(avg.depth = tot.depth / contig.len)

```

### New parse_depth() function

```{r}
cov_df <- mut_depth_df %>%
    group_by(X1) %>%
    summarise(tot.depth = sum(X3)) %>%
    rename("contig.name" = X1) %>%
    left_join(contig_len_df, by = "contig.name") %>%
    mutate(avg.depth = tot.depth / contig.len) %>% 
    arrange(-contig.len)

cov_df


chrom_cov <- cov_df %>%
    filter(contig.len == max(contig.len)) %>%
    pull(avg.depth)

chrom_cov
```


```{r}
library(dplyr)
library(readr)
library(Biostrings)

parse_depth <- function(depth_path, ref_path, min_contig_length=1000, strain, label) {
  # param depth_path: a path to the depth file
  # param ref_path: a path to the reference genome in fasta format
  # min_contig_length: contigs to filter out
  # param strain: strain name
  # param label: mutant or parent
  
  # read reference
  ref_gen <- readDNAStringSet(ref_path)
  # make a table with contig names and lengths
  contig_len_df <- tibble("contig.name" = names(ref_gen), 
                         "contig.len" = width(ref_gen))
  # read depth file
  depth_df <- read.table(depth_path, sep = "\t", header = F)
  # check correct column names
  stopifnot(exprs = {"V1" %in% names(depth_df) & "V3" %in% names(depth_df)})
  # check if the table is empty
  stopifnot(nrow(depth_df) > 0)
  
  # calculate length and avg. coverage of each contig
  cov_df <- depth_df %>%
    group_by(V1) %>%
    summarise(tot.depth = sum(V3)) %>%
    rename("contig.name" = V1) %>%
    full_join(contig_len_df, by = "contig.name") %>%
    mutate(avg.depth = tot.depth / contig.len) %>% 
    arrange(-contig.len) %>% 
    filter(contig.len >= min_contig_length)

  # relative coverage
  chrom_cov <- cov_df %>%
    filter(contig.len == max(contig.len)) %>%
    pull(avg.depth)
  
  cov_df$rel.depth <- cov_df$avg.depth / chrom_cov
  cov_df$strain <- strain
  cov_df$label <- label
  # make a column indicating what is chrom and what is plasmid
  chrom_plasm_col <- c("chrom", rep("plasmid", nrow(cov_df) - 1))
  cov_df$genome.part <- chrom_plasm_col
  
  return(cov_df)
}

# test the function
parse_depth(depth_path = "/home/andrei/Data/HeteroR/results/mutants/amplifications/DA63038/mutant_depth.tsv.gz",
            ref_path = "/home/andrei/Data/HeteroR/results/mutants/variants/DA63038/reference.fasta",
            min_contig_length = 1000,
            strain = "DA63038",
            label = "mutant")

```

```{r}
parse_depth(depth_path = "/home/andrei/Data/HeteroR/results/mutants/copy_number/DA63038/parent_depth.tsv.gz",
            ref_path = "/home/andrei/Data/HeteroR/results/mutants/variants/DA63038/reference.fasta",
            min_contig_length = 1000,
            strain = "DA63038",
            label = "parent")
```



