---
title: "variant calling tests"
author: "AG"
date: "2023-11-21"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Mapping

```{bash}
mamba env create --file var_calling.yaml
```

### Take fasta records from gff annotation file

because they contain correct contig names!

```{bash make.fasta}
GFF="/home/andrei/Data/HeteroR/results/annotations/DA63342/prokka/DA63342_genomic.gff"

# leave fasta, drop the rest
sed -n '/##FASTA/,${p}' $GFF | sed '1d' > assembly_new_names.fasta 
```

```{bash mapping}
READS1="/home/andrei/Data/HeteroR/results/data_filtered/DA63342/Illumina/mutants/DA63342m_1.fq.gz"
READS2="/home/andrei/Data/HeteroR/results/data_filtered/DA63342/Illumina/mutants/DA63342m_2.fq.gz"
REF_FASTA="assembly_new_names.fasta"
# make index first!
# mapping
cd results/variants
conda activate varcalling-env
bowtie2-build $REF_FASTA index
bowtie2 -p 10 -x index -1 $READS1 -2 $READS2 -S aligned.sam && samtools sort -@ 10 aligned.sam > aligned.bam
```

## Variant calling & filtering

```{bash var.calling}
bcftools mpileup --threads 10 --max-depth 800 -f $REF_FASTA -Ou aligned.bam | bcftools call --threads 10 --ploidy 1 -mv -Ob --prior 5.0e-10 -o variants.bcf
bcftools filter -g3 -i 'QUAL>30 && DP>20' -Ob variants.bcf > variants_filtered.bcf
# to take a look
bcftools index variants_filtered.bcf
bcftools query --format "%POS\t%REF\t%ALT\n" variants_filtered.bcf
```

## Variant annotation

```{bash var.annotation}
GFF="/home/andrei/Data/HeteroR/results/annotations/DA63342/prokka/DA63342_genomic.gff"

bcftools view variants_filtered.bcf > variants_filtered.vcf

# drop the genome fasta in the gff file
sed '/##FASTA/,$d' $GFF > DA63342_genomic_clean.gff 

# vcf file must have the same contig names as the gff file
bedtools annotate -i DA63342_genomic_clean.gff -files variants_filtered.vcf > gff_with_annot.tsv
# "Default behavior is to report the fraction of -i covered by each file"
```

```{r read.gff}
library(dplyr)

read_gff <- function(gff_filename) {
  # read and process
  df <-
    as_tibble(read.delim(gff_filename, header = F, comment.char = "#")) %>%
    select(V1, V3, V4, V5, V7, V9, V10) %>%
    filter(V3 == "gene")
  return(df)
}

gff_annotated <- read_gff("/mnt/data/andrei/Data/HeteroR/results/variants/DA63342/DA63342_annotated_variants.gff") %>% filter(V10 > 0)
gff_annotated
```

## Depth and amplifications

```{r, fig.width=14}
library(data.table)

# depth.txt is made by samtools depth/coverage (?)
depth <- fread("/home/andrei/Data/HeteroR/results/variants/DA63150/depth.tsv.gz", sep = "\t")
#depth <- fread("/home/andrei/Data/HeteroR/results/variants/DA63342/depth.tsv.gz", sep = "\t")
# which one is chromosome?
depth %>% group_by(V1) %>% count()
depth_chosen <- depth[V1 == "gnl|UU|CMIAEBAJ_1",]
#depth_chosen <- depth[V1 == "gnl|UU|JBJGLNML_2", ]
depth_chosen
```

### coverage in a window of certain size

calculate z-score

+ 2SD is your candidate

>The absolute value of z represents the distance between that raw score x and the population mean in units of the standard deviation. z is negative when the raw score is below the mean, positive when above.

```{r, fig.width=14, fig.height=4}
#library(dplyr)
library(ggplot2)

window_size <- 1000
mean_coverage <- mean(depth_chosen$V3)

depth_chosen %>%
  mutate(window = 1 + (row_number() - 1) %/% window_size) %>%
  group_by(window) %>%
  summarise(coverage = mean(V3)) %>%
  mutate(z = (coverage - mean(coverage)) / sd(coverage)) %>%
  ggplot(aes(window, z)) +
  geom_point(size = 0.2) +
  geom_hline(yintercept = 2, color = "red")
```


```{r}
depth_per_window <-
        depth_chosen %>%
  mutate(window = 1 + (row_number() - 1) %/% window_size) %>%
  group_by(window) %>%
  summarise(coverage = mean(V3)) %>%
  mutate(z = (coverage - mean(coverage)) / sd(coverage))

depth_per_window %>%
  filter(z >= 3) %>%
        arrange(window)
```

## Overview of the variants

let's check how many and what kind of variants we've got in our 78 strains

First 10 bases in all reads were trimmed - they had very skewed nucleotide content.

```{r, message=FALSE}
library(tidyverse)
library(VariantAnnotation)
library(GenomicRanges)

vcf_files <- dir("/home/andrei/Data/HeteroR/results/variants", pattern = "*.vcf", recursive = T)


var_len <- map_dbl(vcf_files, ~ length(
  readVcf(
    paste0("/home/andrei/Data/HeteroR/results/variants/", .))))

hist(var_len, breaks = 30)
```

Number of variants

```{r}
var_len
```


Turn VCF to tibbles

```{r}
vcf2tibb <- function(file_path, path_prefix){
  # make full path to the file
  full_path <- paste0(path_prefix, file_path)
  # read in vcf
  vcf_df <- as_tibble(rowRanges(readVcf(full_path)))
  # extract strain name: the first 7 characters
  vcf_df$strain <- substr(file_path, 1, 7)
  return(vcf_df)
}

var_df <- map_dfr(vcf_files, ~ vcf2tibb(., path_prefix = "/home/andrei/Data/HeteroR/results/variants/"))

var_df
```

Hist of rows per strain should be the same as the one above

```{r}
var_df %>% 
  group_by(strain) %>% 
  summarise(n = n()) %>% 
  #count() %>% 
  ggplot(aes(n)) +
  geom_histogram(bins = 30)
```

### Variant counts per strain

```{r, fig.width=10, fig.height=3}
var_df %>% 
  group_by(strain) %>% 
  summarise(n = n()) %>% 
  #count() %>% 
  ggplot(aes(strain, n)) +
  geom_col() + 
  theme(axis.text.x = element_text(size = 6, angle = 45, hjust = 1)) +
  geom_hline(yintercept = 3)
```

Ordered by n

```{r}
var_df %>% 
  #filter(strain == "DA63126") %>% 
  group_by(strain, seqnames) %>% 
  dplyr::count() %>% 
  arrange(n)
```

###  Strains with too many variants

```{r}
var_df %>% 
  filter(strain == "DA63126") %>% 
  group_by(seqnames) %>% 
  dplyr::count()
```


```{r}
var_df %>% 
  filter(strain == "DA63784") %>% 
  group_by(seqnames) %>% 
  dplyr::count()
```

```{r}
var_df %>% 
  filter(strain == "DA63870") %>% 
  group_by(seqnames) %>% 
  dplyr::count()
```

#### What I did with DA63126 (and others)

1. fastqc once again: DA63126 has too big GC content deviations. Contamination?
2. fastq_screen
3. Filtering of E.coli mappeing only reads
4. get VCF again (n = 2400, before n = 4000)

Code:

```{bash}
cd results/data_filtered/DA63126/Illumina/mutants
conda activate fastqscreen-env
# tag reads E.coli reads and keep them, toss the rest
fastq_screen --tag --filter 00000001000000 --conf FastQ_Screen_Genomes/fastq_screen.conf DA63126m_1.fq.gz --outdir filter_ecoli
# genome order: hum, mus, rat, dro, worm, yeast, arabido, ecoli, rrna, mito, phix, lamb, vec, adap

# now equalize number of reads in each fastq file
seqkit pair --id-regexp '^(\S+)\/[12]'  -1 filter_ecoli/DA63126m_1.fq.gz -2 filter_ecoli/DA63126m_2.fq.gz

# move the balanced and filtered reads to . and run the pipeline again
# run var_calling.Rmd -> number of variants in DA63126 is 2400, still too much
```


Plots from DA63126 fastq screen:

[screening results reads 1](~/Data/HeteroR/results/data_filtered/DA63126/Illumina/mutants/filter_ecoli/DA63126m_1_screen.png)

[screening results reads 2](~/Data/HeteroR/results/data_filtered/DA63126/Illumina/mutants/filter_ecoli/DA63126m_2_screen.png)

Screening of the other suspicious strains and one of the *clean* strains (DA62954) gives the same looking plots.

From DA63784 fastq screen (200 variants):

[screening results reads 1](~/Data/HeteroR/results/data_filtered/DA63784/Illumina/mutants/filter_ecoli/DA63784m_1_screen.png)


From DA62954 fastq screen (1 variant)

[screening results reads 1](~/Data/HeteroR/results/data_filtered/DA62954/Illumina/mutants/filter_ecoli/DA62954m_1_screen.png)

