---
title: "Feature Engineering"
author: "by A.G."
date: "last update: `r format(Sys.Date(), format = '%d %B %Y')`"
output:
  html_document:
    toc: yes
    df_print: paged
    code_folding: hide
    fig_width: 10
    fig_height: 6
    theme: cerulean
    highlight: kate
    toc_float: yes
  html_notebook:
    df_print: paged
    toc: yes
    number_sections: yes
    toc_float: yes
    theme: cerulean
    highlight: kate
---

```{r, include=F}
knitr::opts_chunk$set(fig.width=10, fig.height=5, message = F, warning = F)
library(dplyr)
library(tidyr)
library(plotly)
library(ggplot2)
library(purrr)
```

This is a notebook for creating various features and their exploration

# Structure of the main table

Which will contain all features and will be suitable for EDA

It should contain the following variables:

-   repeat length,
-   repeat counts,
-   AR length,
-   center spanning,
-   within gene location (of a repeat),
-   distance to oriC,
-   density of matches,
-   density of RG (they often go together)
-   IS presence (?),
-   IS length (=activity)
-   number of plasmids,
-   number of plasmids carrying RGs with repeats,
-   is an RG located on a plasmid,
-   "the black box" of AR

Its tidy structure should look like this:

Each strain contains:

-   [x] 1 AB label

-   [x] 1 resistance label
-   [ ] 1 label for presence of an RG

-   [x] 1 value for plasmid count,

-   [ ] 1 value for a number plasmids carrying RGs,

-   [x] repeat count with min 100
-   [ ] repeat count with min 500
-   [ ] median repeat length with min 100
-   [ ] median repeat length with min 500
-   [ ] maximum repeat length
-   [ ] 1 value for repeat count adjusted by number of plasmids,

-   [x] hundreds of genes; each gene (e.g. region) contains:

    -   [ ] 1 distance to oriC,

    -   [ ] 1 density value,

    -   [ ] 'on a plasmid' label,

    -   [x] 1 value for repeat count,

    -   [ ] 1 value for IS count,

    -   [ ] value for max IS length,

    -   [ ] RGI resistance mechanism,

    -   [ ] RGI resistance SNP,

    -   [ ] RGI other SNP,

    -   [ ] RGI cut-off,

    -   [ ] RGI best identities,

    -   [ ] RGI AMR gene family,

    -   [ ] RGI Model type,

    -   [ ] RGI Drug class,

    -   [ ] RGI ORF ID,

    -   [ ] RGI Percentage Length of Reference Sequence,

    -   [x] a lot of repeat pairs; each repeat pair has:

        -   [x] 1 AR length,
        -   [ ] 1 AR black box vector
        -   [x] 1 repeat length,
        -   [x] 1 center spanning label,
        -   [ ] 'inside the gene' label,

# Making a table with features

## Table with repeats

This one will a basic table for the rest of the analysis.

To build this table anew run the code chunk below (its evaluation is OFF by default)

```{r, eval=FALSE}
# a new way of getting repeat data frame across all strains
strains <- dir("/home/andrei/Data/HeteroR/results/direct_repeats")
repeat_csv_paths <- sapply(strains, function(x){paste0("/home/andrei/Data/HeteroR/results/annotations/",x ,"/repeats/", x, "_repeats.csv")})

repeat_df <- bind_rows(lapply(repeat_csv_paths, function(x){read.csv(x)}))
# calculate AR length
repeat_df$AR_length <- repeat_df$start_2 - repeat_df$end_1

# read bed files to determine center spanning repeat pairs
bed_df <- bind_rows(lapply(strains, function (x){
  filename <- paste0("/home/andrei/Data/HeteroR/results/direct_repeats/", x, "/regions/regions_within.bed")
  df <- read.delim(filename, header=FALSE)
  return(df)
}))

bed_df$gene_center <- round((bed_df$V3 - bed_df$V2 + 1)/2)
bed_df <- rename(bed_df, "record_id"=V4)
bed_df <- select(bed_df, record_id, gene_center)

# join the two tables
repeat_df <- full_join(bed_df, repeat_df, by="record_id")
repeat_df$spans_center <- if_else(repeat_df$end_1 <= repeat_df$gene_center & repeat_df$start_2 >= repeat_df$gene_center, "yes", "no")

```

Or read it from the file *repeat_summary_table.csv* that was made during analysis in `repeat_summary.Rmd`

It contains strains, gene names, repeat coordinates, repeat lengths, AR lengths and 'spans center' label.

Each gene name (aka record_id) is unique.

```{r}
repeat_df <- as_tibble(read.csv("/home/andrei/Data/HeteroR/results/repeats_summary_table.csv")) %>% 
  select(-c(X.1, X)) %>% 
  relocate(strain, .before=record_id)

repeat_df$record_id <- sub("_gene", "", repeat_df$record_id)
```

## Add repeat counts

To each gene - repeats number in region around this gene

To each strain - total repeat count

### Based on all repeats

```{r}
repeat_counts_strain <- repeat_df %>% 
  group_by(strain) %>% 
  summarise("n_repeats_strain"=n())

repeat_counts_gene <- repeat_df %>% 
  group_by(record_id) %>% 
  summarise("n_repeats_gene"=n())

# join 
features <- left_join(repeat_df, repeat_counts_strain, by="strain")
features <- left_join(features, repeat_counts_gene, by="record_id")

features <- rename(features, "repeat_length"=length)

features
```

### Based on repeats > 100 bp

```{r}
repeat_counts_strain_100 <- repeat_df %>% 
  filter(length >= 100) %>% 
  group_by(strain) %>% 
  summarise("n_repeats_strain_min100"=n())

repeat_counts_gene_100 <- repeat_df %>% 
  filter(length >= 100) %>% 
  group_by(record_id) %>% 
  summarise("n_repeats_gene_min100"=n())

# join 
features <- left_join(features, repeat_counts_strain_100, by="strain")
features <- left_join(features, repeat_counts_gene_100, by="record_id")

features
```


### Based on repeats > 500 bp

```{r}
repeat_counts_strain_500 <- repeat_df %>% 
  filter(length >= 500) %>% 
  group_by(strain) %>% 
  summarise("n_repeats_strain_min500"=n())

repeat_counts_gene_500 <- repeat_df %>% 
  filter(length >= 500) %>% 
  group_by(record_id) %>% 
  summarise("n_repeats_gene_min500"=n())

# join 
features <- left_join(features, repeat_counts_strain_500, by="strain")
features <- left_join(features, repeat_counts_gene_500, by="record_id")

features
```


## Add AB testing data

### Read and process AB testing table

```{r}
resistance_df <- as_tibble(read.csv("/home/andrei/Data/HeteroR/resources/resistance_testing.csv", na.strings = ""))
resistance_df <- select(resistance_df, contains("SIR") | contains("strain"))
names(resistance_df) <- sub(".SIR", "", names(resistance_df))
resistance_df
```

### Make a tidy version

```{r}
resistance_df_td <- gather(resistance_df, key="AB", value="resistance", 1:8)
resistance_df_td$AB <- as.factor(resistance_df_td$AB)
resistance_df_td$resistance <- factor(resistance_df_td$resistance, ordered=T, levels = c("S", "I", "R"))
```

I'm not sure which version (regular or tidy) to add to `features`

To use AB and results of testing in plots I need a tidy version

### Add tidy version to features

```{r}
features <- left_join(features, resistance_df_td, by="strain")
```

**OBS!** Now `features` contain `r nrow(features)` rows.

## Number of plasmids

Per strain 

Info on plasmids is in tables `summary.tsv` in `results/assemblies_joined` and it has been collected in `genome_assembly_summary.Rmd`

To build the table anew run the following chunk of code (by default its evaluation is OFF)

```{r, eval=FALSE}
summary_files <- dir(path="/home/andrei/Data/HeteroR/results/assemblies_joined",  pattern = "summary.tsv", recursive=TRUE)
summary_table <- lapply(summary_files, function(x){
  df <- read.delim(paste0("/home/andrei/Data/HeteroR/results/assemblies_joined/", x), stringsAsFactors = TRUE)
  df$Links <- as.integer(gsub(",", "", df$Links, fixed=TRUE))
  df$Length <- as.integer(gsub(",", "", df$Length, fixed=TRUE))
  df$N50 <- as.integer(gsub(",", "", df$N50, fixed=TRUE))
  df$Longest_component <- as.integer(gsub(",", "", df$Longest_component, fixed=TRUE))
  return(df)
  })

summary_table <- bind_rows(summary_table)
```

Or read the table from a file:

```{r, eval=TRUE}
assembly_summary <- as_tibble(read.csv("/home/andrei/Data/HeteroR/results/genome_assembly_summary.csv")) %>% select(-X)

```

Get plasmid counts

```{r}
plasmid_counts <- assembly_summary %>% group_by(Strain, Type) %>% 
  summarise(n_plasmids=n()) %>% 
  filter(Type=="Plasmid") %>% 
  select(-Type) %>% 
  rename("strain"=Strain)
```


and add it to `features`

```{r}
features <- left_join(features, plasmid_counts, by="strain")
# NA in n_plasmids means 0
features$n_plasmids <- ifelse(is.na(features$n_plasmids), 0, features$n_plasmids)
head(features)
```

## Add RGI annotations

### Read RGI annotations 

```{r}
strains <- unique(repeat_df$strain)[!is.na(unique(repeat_df$strain))]

rgi <- bind_rows(lapply(strains, function(x){
  df <- read.delim(paste0("/home/andrei/Data/HeteroR/results/resistance_genes/", x, "/rgi_table.txt"), na.strings = "n/a")
  df$strain <- x
  return(df)
}) )

ab_groups <- unique(rgi$Drug.Class)

# filter rg_df
rgi <- rgi %>% 
  filter( Cut_Off != "Loose" ) %>% 
  select(-c(Predicted_Protein, CARD_Protein_Sequence, Contig, Start, Stop, Orientation, Predicted_DNA, Note))

rgi
```


### Filtering RGI results

**OBS!** Minimum number of RGs found should be much lower than 60 per strain

Criteria:

1. Identity percent

2. Percentage length of reference sequence

3. efflux pumps should go; target alteration should go; permeability should go; inactivation and modification should stay

4. genes with SNPs should go

#### Number of RGs per strain: before filtering

```{r}
n_RG <- rgi %>% group_by(strain) %>% summarize(n_res_genes=n())

ggplot(n_RG, aes(n_res_genes))+ geom_histogram(fill="steelblue")
```

```{r}
summary(n_RG$n_res_genes)
```

These numbers should be lower!

#### Identity percent

For *Strict* hits you may consider *Identity_percent* as another filtering criterion, hits with too low identities should be discarded.

##### Distribution of Identity percentages before filtering

```{r, message=F, fig.width=10, fig.height=5}
hit.ident <-ggplot(rgi %>% select(Cut_Off, Best_Identities), aes(Best_Identities)) +
  geom_histogram(bins = 100, aes(fill=Cut_Off), alpha=0.8) +
  ggtitle("Distribution of identity percentages")

hit.ident
```


##### Leave only 90% and higher

```{r}
rgi_filt <- rgi %>% filter(Best_Identities >= 90)
rgi_filt <- as_tibble(rgi_filt)
summary(rgi_filt$Best_Identities)
```

### Percentage length of reference sequence

Before filtering (except for Best_Identities) it looks like this

```{r, fig.width=10, fig.height=5}
hist.len <-ggplot(rgi_filt %>% select(Cut_Off, Percentage.Length.of.Reference.Sequence), aes(Percentage.Length.of.Reference.Sequence)) +
  geom_histogram(bins = 100, aes(fill=Cut_Off), alpha=0.8) +
  geom_rug(aes(color=Cut_Off))+
  ggtitle("Distribution of percenatge length of reference sequence")

hist.len
```

```{r}
summary(rgi_filt$Percentage.Length.of.Reference.Sequence)
```

How short is too short? 90%

Filtering:

```{r}
rgi_filt <- filter(rgi_filt, Percentage.Length.of.Reference.Sequence >= 90)
summary(rgi_filt$Percentage.Length.of.Reference.Sequence)
```

### Number of RGs per strain : after the 1st filtering

```{r}
n_RG_filt <- rgi_filt %>% group_by(strain) %>% summarize(n_res_genes=n())

ggplot(n_RG_filt, aes(n_res_genes))+ geom_histogram(fill="steelblue")
```
More filtering required!

### Filtering by Resistance.Mechanism

How many resistance mechanisms do we have?

```{r}
unique(rgi_filt$Resistance.Mechanism)
```

What have we decided to filter out?

- efflux

- target alteration

- permeability

What should stay:

- inactivation

- modification

- target replacement?

- target protection?

**OBS!** Leave those records that *contain* what should stay

```{r}
stay <- c("antibiotic target replacement", "antibiotic target protection", "antibiotic inactivation")

rgi_filt <- rgi_filt %>% 
  filter(Resistance.Mechanism %in% stay)

rgi_filt
```

### Filtering SNPs

How many SNPs do we have now?

```{r}
filter(rgi_filt, !is.na(SNPs_in_Best_Hit_ARO))
```

Zero

### Number of RGs per strain: after the 2nd filtering

```{r}
n_RG_filt <- rgi_filt %>% group_by(strain) %>% summarize(n_res_genes=n())

ggplot(n_RG_filt, aes(n_res_genes))+ 
  geom_histogram(fill="steelblue")+
  xlab("Number of RGs per strain")
```

Summary

```{r}
summary(n_RG_filt$n_res_genes)
```

It's OK!

### Add filtered RGI annotations to Features

Add by gene names since they are unique

But first transform them: there should be a column *record_id* like in `features`

```{r}
# make record_id as in features
# transformation example:
# strsplit(rgi_filt$ORF_ID[1], " ")[[1]][1]

rgi_filt$record_id <- map_chr(rgi_filt$ORF_ID, function(x) strsplit(x, " ")[[1]][1])

```


and make a full join with features

Maybe you don't need some of the RGI columns in final `features` df...

```{r}
features <- full_join(features, rgi_filt, by="record_id")
# leave strain.x, remove strain.y
features <- features %>% 
  select(-strain.y) %>% 
  rename("strain"=strain.x)
```

### Remove features of excluded genes

Since many genes were excluded from RGI results but their repeats, counts and AR stay in features,
let's make `features_filtered` without these genes

#### Remove rows in features without Resistance.Mechanism

How many rows don't have resistance mechanisms (i.e. have been filtered outduring RGI filtering?)

```{r}
sum(is.na(features$Resistance.Mechanism))
```


```{r}
features_filtered <- filter(features, !is.na(Resistance.Mechanism))
features_filtered
```

380000 rows instead of 4,5 million


# Plotting features

## Number of plasmids

```{r, fig.width=10, fig.height=4}
ggplot(features %>% filter(!is.na(resistance)) %>% filter(resistance !="I"), aes(AB, n_plasmids)) + 
  geom_boxplot(aes(fill=resistance), notch=T, outlier.alpha = 0.5, outlier.size = 0.5) +
  scale_fill_brewer(palette = "Set2") +
  coord_trans(y="sqrt") +
  xlab("")+
  ylab("n") +
  ggtitle("Plasmid number")
```

## Repeat count

### Based on unfiltered features table

We already know that repeats shorter than 100 mainly make noise

#### Repeats longer than 100 bp

```{r, fig.width=10, fig.height=5}
# prepare df
feat_selected <- features %>%
                   filter(!is.na(resistance)) %>%
                   filter(resistance != "I") %>% 
                   select(strain, AB, resistance, n_repeats_strain_min100, spans_center) %>% 
                   distinct()

ggbox.rc.ab.100 <- ggplot(feat_selected, aes(AB, n_repeats_strain_min100)) +
        geom_boxplot(aes(fill = resistance), notch=F, outlier.alpha = 0.5, outlier.size = 0.5) +
        coord_trans(y="sqrt") +
        facet_grid(spans_center ~ .) +
        scale_fill_brewer(palette = "Set2") +
        ggtitle(label="Repeat count", subtitle="Repeats > 100 bp")

ggbox.rc.ab.100
```

#### Repeats longer than 500 bp

```{r, fig.width=10, fig.height=5}
# prepare df
feat_selected <- features %>%
                   filter(!is.na(resistance)) %>%
                   filter(resistance != "I") %>% 
                   select(strain, AB, resistance, n_repeats_strain_min500, spans_center) %>% 
                   distinct()


ggbox.rc.ab.500 <- ggplot(feat_selected, aes(AB, n_repeats_strain_min500)) +
        geom_boxplot(aes(fill = resistance), notch=F, outlier.alpha = 0.5, outlier.size = 0.5) +
        facet_grid(spans_center ~ .) +
        coord_trans(y="sqrt") +
        scale_fill_brewer(palette = "Set2") +
        ggtitle(label="Repeat count", subtitle="Repeats > 500 bp")

ggbox.rc.ab.500
```

### Based on filtered features table

#### Repeats longer than 100 bp

```{r, fig.width=10, fig.height=5}
# prepare df
feat_selected <- features_filtered %>%
                   filter(!is.na(resistance)) %>%
                   filter(resistance != "I") %>% 
                   select(strain, AB, resistance, n_repeats_strain_min100, spans_center) %>% 
                   distinct()

ggbox.rc.ab.filt.100 <- ggplot(feat_selected, aes(AB, n_repeats_strain_min100)) +
        geom_boxplot(aes(fill = resistance), notch=F, outlier.alpha = 0.5, outlier.size = 0.5) +
        coord_trans(y="sqrt") +
        facet_grid(spans_center ~ .) +
        scale_fill_brewer(palette = "Set2") +
        ggtitle(label="Repeat count", subtitle="Filtered RG, Repeats > 100 bp")

ggbox.rc.ab.filt.100
```

#### Repeats longer than 500 bp

```{r, fig.width=10, fig.height=5}
# prepare df
feat_selected <- features_filtered %>%
                   filter(!is.na(resistance)) %>%
                   filter(resistance != "I") %>% 
                   select(strain, AB, resistance, n_repeats_strain_min500, spans_center) %>% 
                   distinct()

ggbox.rc.ab.filt.500 <- ggplot(feat_selected, aes(AB, n_repeats_strain_min500)) +
        geom_boxplot(aes(fill = resistance), notch=F, outlier.alpha = 0.5, outlier.size = 0.5) +
        coord_trans(y="sqrt") +
        facet_grid(spans_center ~ .) +
        scale_fill_brewer(palette = "Set2") +
        ggtitle(label="Repeat count", subtitle="Filtered RG, Repeats > 500 bp")

ggbox.rc.ab.filt.500
```


## Repeat length

Repeats longer than 100 bp

```{r, fig.width=10, fig.height=5}
# prepare df
feat_selected <- features %>%
                   filter(!is.na(resistance)) %>%
                   filter(resistance != "I") %>% 
                   select(strain, AB, resistance, repeat_length, spans_center) %>% 
                   filter(repeat_length >= 100) %>% 
                   distinct()

ggbox.rl.ab.100 <- ggplot(feat_selected, aes(AB, repeat_length)) +
        geom_boxplot(aes(fill = resistance), notch=F, outlier.alpha = 0.5, outlier.size = 0.5) +
        coord_trans(y="sqrt") +
        facet_grid(spans_center ~ .) +
        scale_fill_brewer(palette = "Set2") +
        ggtitle(label="Repeat length", subtitle="Repeats > 100 bp")

ggbox.rl.ab.100
```

Repeats longer than 500 bp

```{r, fig.width=10, fig.height=5}
# prepare df
feat_selected <- features %>%
                   filter(!is.na(resistance)) %>%
                   filter(resistance != "I") %>% 
                   select(strain, AB, resistance, repeat_length, spans_center) %>% 
                   filter(repeat_length >= 500) %>% 
                   distinct()

ggbox.rl.ab.500 <- ggplot(feat_selected, aes(AB, repeat_length)) +
        geom_boxplot(aes(fill = resistance), notch=F, outlier.alpha = 0.5, outlier.size = 0.5) +
        coord_trans(y="log") +
        facet_grid(spans_center ~ .) +
        scale_fill_brewer(palette = "Set2") +
        ggtitle(label="Repeat length", subtitle="Repeats > 500 bp")

ggbox.rl.ab.500
```


## AR length

For repeats longer than 100 bp

```{r, fig.width=10, fig.height=5}
# prepare df
feat_selected <- features %>%
                   filter(!is.na(resistance)) %>%
                   filter(resistance != "I") %>% 
                   select(strain, AB, resistance, AR_length, repeat_length, spans_center) %>% 
                   filter(repeat_length >= 100) %>% 
                   distinct()

ggbox.arl.ab.100 <- ggplot(feat_selected, aes(AB, AR_length)) +
        geom_boxplot(aes(fill = resistance), notch=F, outlier.alpha = 0.5, outlier.size = 0.5) +
        
        facet_grid(spans_center ~ .) +
        scale_fill_brewer(palette = "Set2") +
        ggtitle(label="AR length", subtitle="Repeats > 100 bp")

ggbox.arl.ab.100

```

------------------------------------------------------------------------

# Save the workspace

```{r, eval=FALSE}
save.image(f="/home/andrei/Data/HeteroR/notebooks/features.RData")
```
