---
title: "Genome assemblies summary"
author: "AG"
date: "last update: `r format(Sys.time(), format = '%d %B %Y, %H:%M CET')`"
output:
  html_document:
    code_folding: hide
    df_print: paged
    toc: true
    toc_float: true
    rmdformats:
      theme: cerulean
      highlight: github
---

Here you'll find a summary of `results/assemblies_joined/{strain}/summary.tsv` files

# Read the data

```{r, message = FALSE, warning = FALSE}
library(dplyr)

# just in case
setwd("/home/andrei/Data/HeteroR/notebooks")

summary_files <- dir(path="../results/assemblies_joined",  pattern = "summary.tsv", recursive=TRUE)
summary_table <- lapply(summary_files, function(x){read.delim(paste0("../results/assemblies_joined/", x), stringsAsFactors = TRUE)})

summary_table <- bind_rows(summary_table)
# replace commas in Length, N50 and Longest component
summary_table$Length <- as.integer(gsub(",", "", summary_table$Length, fixed=TRUE))
summary_table$N50 <- as.integer(gsub(",", "", summary_table$N50, fixed=TRUE))
summary_table$Longest_component <- as.integer(gsub(",", "", summary_table$Longest_component, fixed=TRUE))
summary_table
```


```{r}
summary(summary_table$Type)
```

## Incomplete assemblies

```{r}
incomplete_assemblies <- summary_table %>% filter(Status=="incomplete")
incomplete_assemblies
```

Number of incomplete assemblies (strains) N=`r length(unique(incomplete_assemblies$Strain))`


```{r}
length(unique(summary_table$Strain))
```
```{r}
length(summary_files)
```

No difference

It has existed before

```{r}
summary_files_strains <- sapply(summary_files, function(x){strsplit(x, "/")[[1]][1]})
setdiff(summary_files_strains, summary_table$Strain)
```
These summary files are empty - there are only headers
it happens because I remove first two lines of a summary table in assembly_summary.py
**fixed now**

# Summaries
```{r}
summary(summary_table)
```

## Number of plasmids per strain


```{r, message = FALSE}
plasmids_count <- summary_table %>% group_by(Strain, Type) %>% summarise(Count = n()) %>% filter(Type=="Plasmid")
plasmids_count
```

Total number of strains with plasmids identified

```{r}
length(plasmids_count$Strain)
```

I was unable to find plasmid in 50 strains

```{r, warning = FALSE, message = FALSE, fig.dim = c(16, 8), out.width='150%'}
library(ggplot2)

plot_plasmids_per_strain <- ggplot(plasmids_count, aes(reorder(Strain, -Count), Count)) + geom_col(fill="steelblue") +
        theme(axis.text.x = element_text(size = 6, angle = 90)) +
        xlab("strains") +
        ylab("plasmids count")
plot_plasmids_per_strain
```

```{r, warning = FALSE, message = FALSE, fig.width = 8, fig.asp = 0.6, out.width='100%'}
ggplot(plasmids_count, aes(Type, Count)) +
        geom_boxplot() +
        geom_violin(alpha=0.3) +
        geom_jitter() +
        xlab("") +
        ylab("plasmid count") +
        theme(axis.text.y = element_text(size=15), text = element_text(size=13))
```


```{r}
library(plotly)

fig <- plasmids_count %>%
  plot_ly(
    y = ~Count,
    type = 'violin',
    box = list(
      visible = T
    ),
    meanline = list(
      visible = T
    ),
    x0 = 'Plasmid'
  )

fig <- fig %>%
  layout(
    yaxis = list(
      title = "",
      zeroline = F
    )
  )

fig
```



```{r}
library(plotly)
fig <- plot_ly(x = plasmids_count$Count, type = "histogram")
fig
```

## Number of complete and incomplete segments per strain

```{r, warning = FALSE, message = FALSE}
complete_incomplete <- summary_table %>% group_by(Strain, Status) %>% summarise(Count = n())
complete_incomplete
```
```{r, warning = FALSE, message = FALSE, fig.width = 18, fig.asp = 0.6}
 ggplot(complete_incomplete, aes(reorder(Strain, -Count), Count)) +
          geom_col(aes(fill=Status), position = "stack") +
         theme(axis.text.x = element_text(size = 6, angle = 90),
               axis.text.y = element_text(size = 13), text = element_text(size=13),
               legend.text = element_text(size=16), legend.title = element_blank(), legend.position = c(0.8, 0.8)) +
         xlab("") +
         ylab("number of contigs") +
        scale_fill_brewer(palette="Dark2", labels=c("circular", "linear"))

```

"complete" - circular assembly (Unicycler)

"incomplete" - linear assembly (Unicycler)

"NA" - linear assembly (SPAdes plasmid)


```{r}
summary_table %>% filter(Status=="scaffold")
```

## Genome length distribution

### Chromosomes

```{r, warning = FALSE, message = FALSE}
summary_chrom <- summary_table %>% filter(Type=='Chromosome')
ggplot(summary_chrom, aes(Length)) +
        geom_histogram(stat="bin", fill="#065c2a") +
        theme(axis.text.y = element_text(size=15), text = element_text(size=13), axis.text.x = element_text(size=13)) +
        xlab("Length, bp")

```

### Plasmids

```{r}
summary_plasmids <- summary_table %>% filter(Type=='Plasmid')
ggplot(summary_plasmids, aes(Length)) +
        geom_histogram(stat="bin", fill="#062d5c") +
        theme(axis.text.y = element_text(size=15), text = element_text(size=13), axis.text.x = element_text(size=13)) +
        xlab("Length, bp")

```

These longest outliers (250k, 400k, 700k and 900k) - are they really plasmids?
```{r}
summary_table %>% filter(Type=='Plasmid') %>% arrange(-Length)
```
```{r}
summary_table %>% filter(Strain=="DA63676")
```

Chromosome is incomplete and adds up to 5 Mbp with the longest plasmid...


```{r}
summary_table %>% filter(Strain=="DA63914")
```

Same here...

### Together

```{r}
ggplot(summary_table, aes(Length)) +
        geom_histogram(stat="bin", fill="steelblue", bins = 50) +
        facet_grid(rows=vars(Type)) +
        theme(axis.text.y = element_text(size=15),
              text = element_text(size=13),
              axis.text.x = element_text(size=13),
              legend.title = element_blank(),
              legend.text = element_text(size=16)) +
        xlab("Length, bp") +
        coord_trans(y="sqrt") +
        scale_fill_manual(values=c("#E69F00", "#56B4E9")) +
        scale_x_continuous(breaks=seq(0, 6e+06, 1e+06))
```

# Assembly completeness

Parse BUSCO summary files and summarize the results

```{r}
busco_path <- "/home/andrei/Data/HeteroR/results/qualcheck_assembly/"
busco_filename <- "short_summary.specific.gammaproteobacteria_odb10.busco_results.txt"

busco_files <- dir(path = busco_path, pattern = busco_filename, recursive = TRUE)

busco_list <- lapply(busco_files, function(x){
  strain <- strsplit(x, "/")[[1]][1]
  t <- read.delim(paste0(busco_path, x), skip = 9, header = FALSE, nrows = 6)
  t$Strain <- strain
  return(t)
})

busco_table <- bind_rows(busco_list) %>% select(-c(V1, V4, V5, V6))

busco_table
```

## Distribution of complete BUSCOs

### Complete

```{r}
busco_table %>% filter(V3 == "Complete BUSCOs (C)") %>% summary()
```

### Complete and single-copy

```{r}
busco_table %>% filter(V3 == "Complete and single-copy BUSCOs (S)") %>% summary()
```

### Fragmented

```{r}
busco_table %>% filter(V3 == "Fragmented BUSCOs (F)") %>% summary()
```

```{r}
busco_table %>% filter(V3 == "Fragmented BUSCOs (F)" & V2 > 0)
```

### Missing

```{r}
busco_table %>% filter(V3 == "Missing BUSCOs (MFragmented)") %>% summary()
```

