---
title: "Untitled"
output: html_document
date: "2023-07-04"
---

the alignment file fot the iqtree was initially built over all strains. 
This code is to remove the strains that were not used for modelling. 

find all the strains that were used for the alignment 
```{r}
files <- list.files("/home/michele/HeteroR/results/pangenome/annotations_db/")
files <- files[grep("^DA", files)] 
files <- gsub(".gff", "", files)
```

find list of strains that were used for modelling
```{r}
data_strain <- read_csv("data/features_strain.csv", na = c("NA", "-Inf")) 

# TWO SCHEMES
hr_testing12 <- read_csv("data/heteroresistance_testing.csv", col_select = c(strain, Gr12)) %>% 
  filter(!is.na(strain)) %>% 
  rename("resistance" = Gr12 )

# USE HR 1+2
data_strain <- data_strain %>% 
  left_join(hr_testing12, by = "strain") %>% 
  filter(resistance != "R", !is.na(resistance)) %>% pull(strain)






```

find strains to remove 
```{r}
to_remove <- paste0(files[!files %in% data_strain], "_genomic")
to_remove
```

```{r}
library(Biostrings)
raw_alignment <-readDNAStringSet("/home/michele/HeteroR/results/pangenome/roary_results_gtr/core_gene_alignment.aln")
```

remove

```{r}
clean_alignment <-  raw_alignment[!names(raw_alignment) %in% to_remove]
clean_alignment
```
save
```{r}
writeXStringSet(clean_alignment, '/home/michele/HeteroR/results/pangenome/roary_results_gtr/core_gene_alignment_clean.aln')

```



