---
title: "seq_files_control"
author: "AG"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: yes
    df_print: paged
    code_folding: hide
    fig_width: 10
    fig_height: 6
    theme: cerulean
    highlight: kate
    toc_float: yes
editor_options: 
  markdown: 
    wrap: 72
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = F, warning = F)
#library(purrr)
#library(stringr)
#library(dplyr)
#library(ggplot2)
library(tidyverse)
```


# Getting list of files form Argos

## Number of strains with Illumina dirs

```{bash}
ls ~/Data/Argos/imb_sal_raw/Sequenced_reference_strains/Sequencing/Strains/DA* | grep -c "Illumina"
```

## Number of strains with Nanopore dirs

```{bash}
ls ~/Data/Argos/imb_sal_raw/Sequenced_reference_strains/Sequencing/Strains/DA* | grep -c "Nanopore"
```

## Total number of strains with dirs

```{bash}
ls ~/Data/Argos/imb_sal_raw/Sequenced_reference_strains/Sequencing/Strains | wc -l
```

## List of long reads

###  List of fastq.gz files

```{r}
fastq.gz <- dir(path="~/Data/Argos/imb_sal_raw/Sequenced_reference_strains/Sequencing/Strains/", pattern = "*.fastq.gz", recursive = T)

length(fastq.gz)
```

### List of fastq files

```{r}
fastq <- dir(path="~/Data/Argos/imb_sal_raw/Sequenced_reference_strains/Sequencing/Strains/", pattern = "*.fastq$", recursive = T)

length(fastq)
```

## List of short reads

### List of fq.gz files

```{r}
fq.gz <- dir(path="~/Data/Argos/imb_sal_raw/Sequenced_reference_strains/Sequencing/Strains/", pattern = "*.fq.gz", recursive = T)

length(fq.gz)
```

### List of fq files

```{r}
fq <- dir(path="~/Data/Argos/imb_sal_raw/Sequenced_reference_strains/Sequencing/Strains/", pattern = "*.fq$", recursive = T)

length(fq)
```


## List of strains with at least one fastq.gz or fastq (long reads) file

i.e. strains sequenced with Nanopore

```{r}
nanopore.strains <- unique(c(str_extract(fastq.gz, "DA[0-9]*"), str_extract(fastq, "DA[0-9]*")))

length(nanopore.strains)
```

## List of strains with at least one fq.gz or fq (short reads) file

i.e. strains sequenced with Illumina

```{r}
illumina.strains <- unique(c(str_extract(fq.gz, "DA[0-9]*"), str_extract(fq, "DA[0-9]*")))

length(illumina.strains)
```

## Strain with Nanopore files in the wrong place

It is a strain with Nanopore seqs in Illumina directory

```{bash}
ls ~/Data/Argos/imb_sal_raw/Sequenced_reference_strains/Sequencing/Strains/DA63104/Illumina/
```

But it has normal Illumina reads as well

## Strains with both long and short reads

```{r}
length(intersect(illumina.strains, nanopore.strains))
```

## Strains with long reads, but without short reads

```{r}
setdiff(nanopore.strains, illumina.strains)
```

## Strains with short reads but without long reads

```{r}
setdiff(illumina.strains, nanopore.strains)
```

These are the strains we need to sequence

## Write strain names to files

```{r}
write(nanopore.strains, "~/Data/HeteroR/resources/strain_lists/nanopore_sequenced_strains.txt", ncolumns = 1)

write(illumina.strains, "~/Data/HeteroR/resources/strain_lists/illumina_sequenced_strains.txt", ncolumns = 1)

write(intersect(nanopore.strains, illumina.strains), "~/Data/HeteroR/resources/strain_lists/completely_sequenced_strains.txt", ncolumns = 1)

write(setdiff(illumina.strains, nanopore.strains), "~/Data/HeteroR/resources/strain_lists/illumina_only_strains.txt", ncolumns = 1)

write(setdiff(nanopore.strains, illumina.strains), "~/Data/HeteroR/resources/strain_lists/nanopore_only_strains.txt", ncolumns = 1)
```


# Nanopore coverage

## Get coverage data

Get table with coverage using your python script

```{bash, eval=FALSE}
# run from ~/Data/HeteroR

python workflow/scripts/coverage.py nanopore_sequenced_strains.txt 5131220 results/coverage/nanopore_sequenced_strains_raw.tsv raw

python workflow/scripts/coverage.py nanopore_sequenced_strains.txt 5131220 results/coverage/nanopore_sequenced_strains_raw.tsv filtered

```

Read the results and add completeness information

## Read raw coverage data

```{r}
nano_cov_raw <- read_delim("~/Data/HeteroR/results/coverage/nanopore_sequenced_strains_raw.tsv")
nano_cov_raw$strain <- str_extract(nano_cov_raw$file, "DA[0-9]*")
nano_cov_raw <- relocate(nano_cov_raw, strain, .before = "file")
nano_cov_raw$filtered <- "raw"

nano_cov_raw
```

## Read filtered coverage data

```{r}
nano_cov_filtered <- read_delim("~/Data/HeteroR/results/coverage/nanopore_sequenced_strains_filtered.tsv")
nano_cov_filtered$strain <- str_extract(nano_cov_filtered$file, "DA[0-9]*")
nano_cov_filtered <- relocate(nano_cov_filtered, strain, .before = "file")
nano_cov_filtered$filtered <- "filt"

nano_cov_filtered
```

## Join them together

```{r}
nano_cov <- bind_rows(nano_cov_filtered, nano_cov_raw)

nano_cov
```

## Add chromosome completeness data

```{r}
assemblies_summary <- read_csv("~/Data/HeteroR/results/tables/genome_assembly_summary.csv")
assemblies_summary <- assemblies_summary %>% 
  select(Length, N50, Longest_component, Status, Status2, Strain, Type) %>% 
  filter(Type == "Chromosome") %>% 
  distinct() %>% 
  rename(strain="Strain")

assemblies_summary
```

```{r}
nano_cov <- left_join(nano_cov, assemblies_summary, by="strain")

nano_cov
```

## Plot

```{r}
ggplot(nano_cov %>% filter(!is.na(Status)), aes(filtered, coverage))+
  geom_boxplot(aes(fill=Status), varwidth = T, notch = T)+
  coord_trans(y="sqrt")
```

# Write strains that should be resequenced to file

```{r}
nano_cov %>% filter(filtered=="filt", Status == "incomplete") %>% arrange(coverage)
```


```{r}
strains_incomplete_filt <- nano_cov %>% 
  filter(filtered=="filt", Status == "incomplete") %>% 
  arrange(coverage) %>% 
  select(strain, coverage)

strains_incomplete_raw <- nano_cov %>% 
  filter(filtered=="raw", Status == "incomplete") %>% 
  arrange(coverage) %>% 
  select(strain, coverage)

strains_incomplete <- full_join(strains_incomplete_filt, strains_incomplete_raw, by="strain")

write.csv(strains_incomplete, "~/Data/HeteroR/results/tables/incomplete_assemblies.csv", row.names = F)
```

# Predict completeness of assembly

Attempt number 2, on a bigger data set.

Add raw data statistics to assemblies summary

## Get data

```{r}
library(caret)

my_data <- left_join(
  select(assemblies_summary, c(strain, Status)),
  select(nano_cov_raw, c(strain, num_seqs, sum_len, min_len, avg_len, max_len, coverage)),
  by="strain"
  )

my_data$Status <- as.factor(my_data$Status)
#my_data$sum_len_log <- log(my_data$sum_len)

summary(my_data)
```

## Make tidy

```{r}
my_data_td <- my_data  %>% gather(key="statistic", value="value", 3:8)
```

## Box plots

```{r, fig.width=10}
ggplot(my_data_td, aes(Status, value))+
  geom_boxplot(varwidth = T, aes(fill=Status))+
  coord_trans(y="log")+
  facet_grid(. ~ statistic)+
  scale_fill_brewer(palette = "Set1")+
  xlab("")+
  ylab("")
```

## Pairs plot

```{r, fig.width=10}
library(GGally)
ggpairs(my_data, 
        columns=c(3:8), 
        aes(color=Status, alpha=0.2, dotsize=0.02), 
        upper = list(continuous = wrap("cor", size = 2.5)),
        diag=list(continuous ="barDiag"))+
  scale_color_brewer(palette = "Set1")+
  scale_fill_brewer(palette = "Set1")

```

## Model

```{r}
set.seed(100)

inTrain <- createDataPartition(y = my_data$Status, p=0.7, list=FALSE)

training <- my_data[inTrain,]
testing <- my_data[-inTrain,]

```

